pkg/k8s/resource/diffstore.go

```diff
@@ -0,0 +1,270 @@
+		select {
+		case event, ok := <-updateChan:
+			if !ok {
+				break
+			}
+
+			event.Handle(
+				func() error {
+					dsf.inSync = true
+					dsf.signalObservers()
+					return nil
+				},
+				dsf.handleChange,
+				dsf.handleChange,
+			)
+		case err := <-errChan:
+			dsf.params.Shutdowner.Shutdown(hive.ShutdownWithError(err))
+			return
+		}
+	}
+}
+
+func (dsf *diffStoreFactory[T]) handleChange(k Key, _ T) error {
+	dsf.mu.Lock()
+	defer dsf.mu.Unlock()
+
+	for _, store := range dsf.diffStores {
+		store.markChanged(k)
+	}
+
+	if dsf.inSync {
+		dsf.signalObservers()
+	}
+
+	return nil
+}
+
+func (dsf *diffStoreFactory[T]) signalObservers() {
+	for _, onDiff := range dsf.onDiff {
+		select {
+		case onDiff <- struct{}{}:
+		default:
+		}
+	}
+}
+
+// DiffStore is a super set of the Store. The diffStore tracks all changes made to the diff store since the
+// last time the user synced up. This allows a user to get a list of just the changed objects while still being able
+// to query the full store for a full sync.
+type DiffStore[T k8sRuntime.Object] interface {
+	Store[T]
+
+	// Diff returns a list of items that have been upserted(updated or inserted) and deleted since the last call to Diff.
+	Diff() (upserted []T, deleted []Key, err error)
+	// MarkChanged marks a key as having been changed so it will appear in the Diff again. This can be used in case
+	// of a failure to process so a next call to Diff gets the chance to process the key again.
+	MarkChanged(Key)
+}
+
+var _ DiffStore[*k8sRuntime.Unknown] = (*diffStore[*k8sRuntime.Unknown])(nil)
+
+// diffStore takes a Resource[T] and watches for events, it stores all of the keys that have been changed.
+// diffStore can still be used as a normal store, but adds the Diff function to get a Diff of all changes.
+// The diffStore also takes in Signaler which it will signal after the initial sync and every update thereafter.
+type diffStore[T k8sRuntime.Object] struct {
+	store Store[T]
+
+	factory *diffStoreFactory[T]
+
+	mu          lock.Mutex
+	changedKeys map[Key]bool
+}
+
+// Diff returns a list of items that have been upserted(updated or inserted) and deleted since the last call to Diff.
+func (sd *diffStore[T]) Diff() (upserted []T, deleted []Key, err error) {
+	sd.mu.Lock()
+	defer sd.mu.Unlock()
+
+	for k := range sd.changedKeys {
+		item, found, err := sd.store.GetByKey(k)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if found {
+			upserted = append(upserted, item)
+		} else {
+			deleted = append(deleted, k)
+		}
+	}
+
+	sd.changedKeys = make(map[Key]bool, 0)
+
+	return upserted, deleted, err
+}
+
+// MarkChanged marks a key as having been changed so it will appear in the Diff again. This can be used in case
+// of a failure to process so a next call to Diff gets the chance to process the key again.
+func (sd *diffStore[T]) MarkChanged(key Key) {
+	sd.markChanged(key)
+
+	// Signal all observers to re-trigger an observer activated Diff.
+	sd.factory.signalObservers()
```

Comment:
we're not holding the right lock when calling `signalObservers()` (`dsf.onDiff` unprotected)